Revision to leap seconds and calendar conventions (#604)
cf-convention/cf-conventions: ch04.adoc

The `computed_standard_name` attribute indicates that the values in variable
`p` would have a `standard_name` of `air_pressure`.

[[time-coordinate, Section 4.4, "Time Coordinate"]]
=== Time Coordinate

Variables representing time axes have time coordinates as their elements.

> In CF data, a time coordinate is a number which identifies an instant along the continuous axis of time, expressed as elapsed time relative to a specified reference point.

Author: I take the point that "In CF data" is unnecessary. Instead of "point in time" I have put "instant", which is a word we have used quite a lot in 4.4, apparently without problems.

Variables containing time coordinates have a **`units`** attribute (see <<time-coordinate-variables>> and <<time-coordinate-units>>, e.g.\ **`units="days since 1990-1-1 0:0:0"`**) which defines both the unit of measure of elapsed time (here **`days`**) and the reference point in time (here **`1990-1-1 0:0:0`**).

The interpretation of the numeric values as dates and times depends on both the **`units`** attribute and the calendar, specified by the **`calendar`** attribute.
Given the **`units`** and **`calendar`**, a time coordinate value and its corresponding representation as a date and time are interconvertible.

> CF defines several different calendars (see~<<calendar>>) to meet diverse requirements, both for the real world and for models.

Reviewer: The `~` renders
```suggestion
CF defines several different calendars (see <<calendar>>) to meet diverse requirements, both for the real world and for models.
```

In reality, the need for explicit calendar definitions arises in part because the natural astronomical periods of year and day, which are based on the Earth's orbit and rotation, are not constants and do not align exactly.
A year is not an integer multiple of days, the length of a day varies with changes in Earth's rotation, and the timing of seasons is affected by the precession of the equinoxes.

> Calendars introduce an extra day (a "leap day", in leap years) to keep annual cycles aligned with the seasons, and leap seconds are occasionally inserted into Coordinated Universal Time (link:$$https://en.wikipedia.org/wiki/Coordinated_Universal_Time$$[UTC]) to maintain agreement with observed rotation.

Reviewer 2: I like this suggestion.

> Calendars introduce an extra day (a "leap day", in leap years) to keep annual cycles aligned with the seasons, and leap seconds are occasionally inserted into Coordinated Universal Time (link:$$https://en.wikipedia.org/wiki/Coordinated_Universal_Time$$[UTC]) to maintain agreement with observed rotation.

Author: I have adopted "astronomical year" and put "to keep their annual cycles aligned" i.e. the calendars' cycles.

See <<leap-seconds>>.

The CF standard follows UDUNITS (<<units>>) in the definition of the acceptable
units of measure for time.
The most commonly used of these units strings (and their abbreviations) are **`day`** (**`d`**), **`hour`** (**`hr`**, **`h`**), **`minute`** (**`min`**) and **`second`** (**`sec`**, **`s`**).
Plural forms are also acceptable.

UDUNITS defines a **`year`** to be exactly 365.242198781 days (the interval between 2 successive passages of the sun through vernal equinox).
__It is not a calendar year.__ UDUNITS defines a  **`month`** to be exactly **`year/12`**, which is __not a calendar month__.
We recommend that **`year`** and **`month`** should not be used, because of the potential for mistakes and confusion.

UDUNITS defines a **`minute`** as 60 **`seconds`**, an **`hour`** as 3600 **`seconds`** and a **`day`** as 86400 **`seconds`**.
These are not calendar units.
When a leap second is inserted into UTC, the minute, hour and day affected differ by one second from their usual durations according to clock time, but the units of **`minute`**, **`hour`** and **`day`** do not; they are fixed units of measure.

> To avoid mistakes and confusion, we therefore recommend that these units should not be used in the **`utc`** calendar; instead, we recommend the **`second`** as the unit for time coordinates in that calendar.

Reviewer: Further to the comments on this for the Conformance document, I'd like to suggest adding  "submultiples of second" here. These are often used in L0/L1 remote sensing datasets.

```suggestion
To avoid mistakes and confusion, we therefore recommend that these units should not be used in the **`utc`** calendar; instead, we recommend the **`second`**, or any submultiple of **`second`** defined in <<table-supported-units>>, as the unit for time coordinates in that calendar.
```

Reviewer 2: @JonathanGregory: SI specifically rules out [sub-]multiples with minute, hour, day:
"The SI prefixes can be used with several of these units, but not, for example, with the non-SI units of time."
(from section 4 as cited above). They are OK for second, and year (and any other more or less esoteric time unit of measure) .

@ChrisBarker-NOAA: Well as far as UDUNITS "understand" any unit, multiples and submultiples of all time units are fine, here are a few examples:
```bash
$ udunits2 -H "s" -W "ns"
    1 s = 1e+09 ns
    x/ns = 1e+09*(x/s)
$ udunits2 -H "Ps" -W "ns"
    1 Ps = 1e+24 ns
    x/ns = 1e+24*(x/Ps)
$ udunits2 -H "Ps" -W "year"
    1 Ps = 3.16888e+07 year
    x/year = 3.16888e+07*(x/Ps)
$ udunits2 -H "nyr" -W "ms"
    1 nyr = 31.5569 ms
    x/ms = 31.5569*(x/nyr)
```
As you suggest, I think it is more limitations in other software.

For compatibility with other software, CF strongly recommends that `since` should be used.

The reference datetime string (appearing after the identifier **`since`**) is required.
It must include the date, which may be followed by time or time zone offset or both.
Its format is __y__-__m__-__d__ [__H__:__M__:__S__] [__Z__], where [...] indicates an optional element, as follows:

* _y_ is year, _m_ month, _d_ day, _H_ hour and _M_ minute, which are all integers of one or more digits, and _y_ may be prefixed with a sign (but note that some CF calendars do not permit negative years; see <<calendar>>),

* _S_ is second, which may be integer or floating point (see <<leap-seconds>> regarding __S__>59),

* _Z_ is the time zone offset.
_Z_ is _not_ a time zone name or acronym; it is an interval of time, specified in one of the formats described below.
The default time zone offset is zero.
We recommend that a non-zero time zone offset should not be specified, because it is easy to make mistakes about the sign of the offset and allowance for daylight-saving/summer time.
A non-zero offset is _not allowed_ in the **`utc`** and **`tai`** calendars (see <<calendar>>).

> While the default is an offset of zero, we suggest that a zero offset be stated explicitly to avoid any confusion where omitting it might be misunderstood as indicating local time.

Reviewer: Just to say that I originally thought "suggest" should be replaced by a BCP14-aligned word, but then I realised that we can't check this, so "suggest" is just fine, I suggest!

In a time zone with zero offset, time (approximately) equals mean solar time for 0 `degrees_east` of longitude.
(Although this may be exact in a model, in reality the time with zero time zone offset differs by some seconds from mean solar time; see the discussion of UTC and leap seconds in <<calendar>>.)
If both time and time zone offset are omitted the time is 00:00:00 (the beginning of the day i.e. midnight at 0 `degrees_east`).
Thus, **`units = "days since 1990-1-1"`** means the same as **`units = "days since 1990-1-1 00:00:00"`**.

> Only numbers (digits, `+`, `-` and `:`) or the letter "Z" are allowed in _Z_.

Author: Yes, I agree. I have rearranged the text about _Z_, I hope to make it clearer and less repetitive.

> Only numbers (digits, `+`, `-` and `:`) or the letter "Z" are allowed in _Z_.

Reviewer: Thanks - I like your new version.

> The **`units`** attribute of a time coordinate variable takes a string value that follows the formatting requirements of the <<UDUNITS>> package (e.g. <<time-axis-ex>>).

Reviewer 2: I agree with @davidhassell

> The **`units`** attribute of a time coordinate variable takes a string value that follows the formatting requirements of the <<UDUNITS>> package (e.g. <<time-axis-ex>>).

Author: I have kept the reference.

It must comprise a unit of measure that is physically equivalent to the SI base unit of time (i.e. the second), followed by the word **`since`** and a reference datetime.
The format of the **`units`** string implies that the time coordinate equals the length of the time interval from the instant identified by the reference datetime to the instant identified by the time coordinate.
This is exactly true in all cases except when leap seconds occur between the two intervals in the **`standard`** calendar.

> See <<leap-seconds>>.

Author: I agree it's not necessary for understanding the statement, but this is a point that has often been controversial or problematic. As Lars says, it's quite possible that the reader will want to know more if they are using the `standard` calendar. The reference tells them where to look if they are curious. I have changed the reference to `(See <<leap-seconds>> for details.)`.

> The CF standard follows UDUNITS (<<units>>) in the definition of the acceptable

Author: OK. I have made a note of this for discussion in another issue.

These are not calendar units.
When a leap second is inserted into UTC, the minute, hour and day affected differ by one second from their usual durations according to clock time, but the units of **`minute`**, **`hour`** and **`day`** do not; they are fixed units of measure.
To avoid mistakes and confusion, we therefore recommend that these units should not be used in the **`utc`** calendar; instead, we recommend the **`second`** as the unit for time coordinates in that calendar.
See also <<calendar>> and <<leap-seconds>>.

UDUNITS permits a number of alternatives to the word **`since`** in the units of time coordinates.
All the alternatives have exactly the same meaning in UDUNITS.
For compatibility with other software, CF strongly recommends that `since` should be used.

The reference datetime string (appearing after the identifier **`since`**) is required.
It must include the date, which may be followed by time or time zone offset or both.
Its format is __y__-__m__-__d__ [__H__:__M__:__S__] [__Z__], where [...] indicates an optional element, as follows:

> * _y_ is year, _m_ month, _d_ day, _H_ hour and _M_ minute, which are all integers of one or more digits, and _y_ may be prefixed with a sign (but note that some CF calendars do not permit negative years; see <<calendar>>),

Author: Yes, that's true, but I think it's worth saying in case the reader is worried by or curious about "may be prefixed with a sign". They might say to themselves, "Oh - are negative years allowed, then? Does -1 mean 2 BC or 1 BC?" or other such concerns. The reference tells them where to look. People don't necessarily read the document from one end to the other. They might alight anywhere in it.

> * _S_ is second, which may be integer or floating point (see <<leap-seconds>> regarding __S__>59),

Author: The text now says `(see <<leap-seconds>> regarding _S_)` without `>59`. Like the issue of negative years, the cross-reference is to tell the reader where to look concerning the values of _S_. I've shortened it further to `(see <<leap-seconds>>)`. I think it's obvious what the reference is about, since this will be expanded to `(see Section 4.4.4, "Leap Seconds")`.

* _Z_ is the time zone offset.
_Z_ is _not_ a time zone name or acronym; it is an interval of time.

The default for time zone offset _Z_ is zero, which may also be explicitly indicated in any of the numeric formats for _Z_ defined below, or by the letter `Z`, sometimes referred to as "Zulu Time".
We suggest that a zero offset be stated explicitly to avoid confusion in any situation where omitting it might be misunderstood as indicating local time.

> In a time zone with zero offset, time (approximately) equals mean solar time for 0 `degrees_east` of longitude.

Author: While of course it's true that, "In a time zone with zero offset, time equals UTC," in the `utc` calendar and the `standard` calendar applied to the real world since 1972, it's not a meaningful statement in all the other situations, where UTC Is not defined. In response to this comment, I've moved a couple of sentences, in order to gather together the statements about time zone offsets. This change has the effect of stating that UTC must have zero offset immediately before the sentence you're commenting on, thus:

> ... We suggest that a zero offset be stated explicitly to avoid confusion in situations where omitting it might be misunderstood as indicating local time. We recommend that a non-zero time zone offset should _not_ be specified _in any situation_, because it is easy to make mistakes about the sign of the offset and allowance for daylight-saving/summer time. A non-zero offset is _not allowed_ in the **`utc`** and **`tai`** calendars (see `<<calendar>>`).

> In a time zone with zero offset, time (approximately) equals mean solar time for 0 **`degrees_east`** of longitude. ...

I hope that makes clearer that this sentence is intended to explain what zero offset means in a physical way. I inserted the italics in order to draw attention to the differences between the first and second sentences quoted, which might be missed by the reader.

> (Although this may be exact in a model, in reality the time with zero time zone offset differs by some seconds from mean solar time; see the discussion of UTC and leap seconds in <<calendar>>.)

Reviewer 2: I agree with @pvanlaake.

If both time and time zone offset are omitted the time is 00:00:00 (the beginning of the day i.e. midnight at 0 `degrees_east`).

> Thus, **`units = "days since 1990-1-1"`** means the same as **`units = "days since 1990-1-1 00:00:00"`**.

Author: I haven't changed it because this is what we have in CF 1.12, and in the majority (I believe, without counting) of similar statements in the document. I have noted it for discussion as a general issue.

> Thus, **`units = "days since 1990-1-1"`** means the same as **`units = "days since 1990-1-1 00:00:00"`**.

Reviewer 3: @pvanlaake -- I'm in agreement, but I also agree with @larsbarring that we should make it a separate issue.  When the time comes, I will be happy to help with the editing!

> We recommend that a non-zero time zone offset should not be specified, because it is easy to make mistakes about the sign of the offset and allowance for daylight-saving/summer time.

Author: CF 1.12 contains 11 instances of "We recommend" and 17 of "It is recommended", so I don't think we need to change this one in particular. There's a general issue about whether we might change all such statements to conform with BCP14. That's a different open issue #546.

A non-zero offset is _not allowed_ in the **`utc`** and **`tai`** calendars (see <<calendar>>).

Other than `Z` for zero offset, the time zone offset _Z_ must be in one of the following four numeric formats, each of which may optionally be prefixed by a positive sign `+` or negative sign `-`, the default sign being positive:

** _H_, the hour alone, of one or two digits e.g. **`-6`**, **`2`**, **`+11`**, which is sufficient for many time zones.

** __H__:__M__, where _H_ is hour and _M_ minute, each of one or two digits, and separated by a colon `:`, e.g. **`5:30`**.

** four digits, of which the first pair are the hours and the second the minutes e.g. **`0530`**.

** three digits, of which the first is the hour (0--9) e.g. **`530`**.

Spaces are not allowed in the numeric formats.
If _Z_ is `Z` or begins with `+` or `-`, the space before _Z_ may be omitted.

> For example, **`seconds since 1992-10-8 15:15:42.5 -6:00`** indicates seconds since October 8th, 1992 at 3 hours, 15 minutes and 42.5 seconds in the afternoon, in a time zone where the datetime is six hours behind the default.

Author: I don't think we should refer to UTC, because that's applicable only for the `utc` calendar and the `standard` calendar after 1972. I agree that "behind the default" is not clear. Possibly "earlier than" might be confusing too, since the two datetimes refer to the same instant, so neither is earlier. Since you draw attention to it, I think it's clearer if we just state the two equivalent datetimes, without describing the relationship between them, to illustrate the point. Also, we can simplify it by omitting `seconds since`, because this part of the text is just about the datetime format, and by using the same example in both sentences. I think we can make it easier to read by not describing the time of day so laboriously. This I've replace this sentence and the next one with:

> For example, **`1992-10-8 15:15:42.5`** indicates 42.5 seconds after 3.15 pm on 8th October 1992, in the time zone with zero offset. Subtracting the time zone offset from a given datetime converts it to the equivalent datetime with zero time zone offset e.g. **`1992-10-8 09:15:42.5 -6`** identifies the same instant as **`1992-10-8 15:15:42.5`**.

We recommend that **`year`** and **`month`** should not be used, because of the potential for mistakes and confusion.

UDUNITS defines a **`minute`** as 60 **`seconds`**, an **`hour`** as 3600 **`seconds`** and a **`day`** as 86400 **`seconds`**, consistent with the <<SI>> definitions of these non-SI units.
These are fixed units of measure.
When a leap second is inserted into UTC, the minute, hour and day affected differ by one second from their usual durations according to clock time, but the units of **`minute`**, **`hour`** and **`day`** do not.
To avoid mistakes and confusion, we therefore recommend that these units should not be used in the **`utc`** calendar.
See also <<calendar>> and <<leap-seconds>>.

UDUNITS permits a number of alternatives to the word **`since`** in the units of time coordinates.
All the alternatives have exactly the same meaning in UDUNITS.
For compatibility with other software, CF strongly recommends that **`since`** should be used.

The reference datetime string (appearing after the identifier **`since`**) is required.
Its format is __y__-__m__-__d__ [__H__:__M__:__S__[ __T__]], where [...] indicates an optional element, as follows:

> * __y__ is year, __m__ month, __d__ day, __H__ hour and __M__ minute, which are all integers of one or more digits, and __y__ may be prefixed with a sign (but note that some CF calendars do not permit negative years; see <<calendar>>),

Reviewer 2: ```suggestion
* __y__ is year, __m__ month, __d__ day, __H__ hour and __M__ minute, __y__ is one or more digits and may be prefixed with a sign (but note that some CF calendars do not permit negative years; see <<calendar>>), all other are one or two digits and must not have a sign.

We suggest that a zero offset be stated explicitly to avoid confusion in situations where omitting it might be misunderstood as indicating local time.
We recommend that a non-zero time zone offset should __not__ be specified __in any situation__, because it is easy to make mistakes about the sign of the offset and allowance for daylight-saving/summer time.
A non-zero offset is __not allowed__ in the **`utc`** and **`tai`** calendars (see <<calendar>>).

In a time zone with zero offset, time (approximately) equals mean solar time for 0 **`degrees_east`** of longitude.
If both time and time zone offset are omitted the time is 00:00:00 (the beginning of the day i.e. midnight in the time zone with zero offset).
Thus, **`units = "days since 1990-1-1"`** means the same as **`units = "days since 1990-1-1 00:00:00Z"`**.

Other than **`Z`** for zero offset, the time zone offset __T__ must be in one of the following four signed numeric formats, where &pm; stands for the positive sign **`+`** or the negative sign **`-`**, and __n__ stands for a single digit:

> * &pm;__H__, the hour alone, of one or two digits e.g. **`-6`**, **`-0`**, **`+0`**, **`+11`**, which is sufficient for many time zones.

Reviewer 2: ```suggestion
* &pm;__H__, the hour alone, of one or two digits e.g. **`-6`**,**`+0`**, **`+11`**, which is sufficient for many time zones.```

We must not allow `-0` or `-00`!!  It is not included  in widely used and generally accepted standards (ISO), and in some it has a special meaning (something like "it's UTC time with unknow timezone"). Currently UDUNITS2 accepts this, which must not be used as an argument because its datetime handling is pretty fragile and buggy (I am currently doing some work on this).

**`standard`**:: The Gregorian--Julian calendar as defined by UDUNITS.
This is the default calendar when there is no **`calendar`** attribute, except when the attribute `month_lengths` is present (see <<explicit-calendar>>).
A deprecated alternative name for this calendar is **`gregorian`**.
The Gregorian and Julian calendars have the same lengths of their months; they differ only in respect of the rules that decide which years are leap years.
In the **`standard`** calendar, datetimes after and including 1582-10-15 0:0:0 are in the Gregorian calendar, in which a year is a leap year if either (i) it is divisible by 4 but not by 100 or (ii) it is divisible by 400.
Datetimes before (and excluding) 1582-10-5 0:0:0 are in the Julian calendar, in which any year that is divisible by 4 is a leap year.
Year 1 AD or CE in the **`standard`** calendar is also year 1 of the **`julian`** calendar.
Negative years are invalid in time coordinates and reference datetimes in the **`standard`** calendar.
+
In the **`standard`** calendar, datetimes in the range from (and including) 1582-10-5 0:0:0 until (but excluding) 1582-10-15 0:0:0 are invalid.

> Datetimes in this range must not be used as reference in **`units`**, and the interval between the datetimes 1582-10-15 0:0:0 and 1582-10-4 0:0:0 is exactly 1 day.

Reviewer 3: I was confused for a long moment by the fact that 10-15 is first and 10-04 is second.  I would suggest swapping them to the natural order: `interval between the datetimes 1582-10-4 0:0:0 and 1582-10-15 0:0:0`

Author: Done. I can't remember why it was backwards!

It is recommended that a reference datetime before the discontinuity should not be used for datetimes after the discontinuity, and vice-versa.
+
__The **`standard`** calendar is used both for (mostly observational) data with real-world datetimes, including leap seconds (after 1972), and for data not including leap seconds__, generated by models according to the Gregorian calendar.

> Therefore, __when using data written with the **`standard`** calendar, you should regard datetimes and time coordinates as uncertain by a number of seconds (but less than a minute),__ because leap seconds may have been counted when you expected them to have been ignored, or ignored when you expected them to have been counted.

Reviewer 2: ```suggestion
Therefore, when using data written with the **`standard`** calendar the datetimes and time coordinates may be uncertain by a number of seconds (but less than a minute)  because it is unknown whether leap seconds have been taken into account or not.
```

For producing new datasets, where precision to the second is important, it is recommended to use the **`utc`** calendar for data (whether observed or simulated) that follows the real-world calendar according to UTC including leap seconds, and the **`proleptic_gregorian`** calendar for model-generated data that follows the Gregorian calendar without leap seconds.

**`proleptic_gregorian`**:: A calendar with the Gregorian rules for leap years extended to dates before 1582-10-15.
All dates consistent with these rules are allowed, both before and after 1582-10-15 0:0:0.

**`julian`**:: Julian calendar, in which a year is a leap year if it is divisible by 4, even if it is also divisible by 100.
Year 1 AD or CE in the **`julian`** calendar is also year 1 of the **`standard`** calendar.
Negative years are invalid in time coordinates and reference datetimes in the **`julian`** calendar, and leap seconds are not allowed.

**`utc`**:: A Gregorian calendar __with__ leap seconds as prescribed by UTC.
The current definition of UTC came into force at 1972-01-01 00:00.

> This instant has the datetime 1972-01-01 00:00:10 (10 seconds greater) in the **`tai`** calendar.

Reviewer 2: ```suggestion
This instant has the datetime 1972-01-01 00:00:10 in the **`tai`** calendar. This 10 second difference was accumulated durng the period from 1958-01-01 when fractional seconds occasionally were added to to keep the predecessor to UTC in phase with earth rotation.

Earlier datetimes than 1972-01-01 00:00 are not allowed in the **`utc`** calendar.
For any given instant, the **`utc`** datetime is behind the **`tai`** datetime, where "behind" means the same as it does when describing a timezone to the west as being behind one to the east.
Moreover, a given datetime in the **`utc`** calendar represents an instant that is earlier than the same datetime in the **`tai`** calendar.

> The difference between the two datetimes (**`tai`** minus **`utc`**) for a given instant of time is the net number of leap seconds introduced since 1958-01-01.

Reviewer 4: ```suggestion
The difference between the two datetimes (**`tai`** minus **`utc`**) for a given instant of time is 10 seconds plus the net number of leap seconds introduced since 1972-01-01.
```

Author: Thanks, Patrick. I have put it in.

When a datetime is converted to a time coordinate value or vice-versa in this calendar, any leap seconds (positive or negative) must be counted that occurred in the interval between the datetime and the reference datetime in the **`units`**.
Datetimes in the future are not allowed in this calendar, because it is unknown when future leap seconds will occur.
A non-zero time zone offset is not allowed in this calendar.

**`tai`**:: A Gregorian calendar __without__ leap seconds that is based on International Atomic Time (TAI), which began on 1958-01-01 0:0:0.
Earlier datetimes than this are not allowed in the **`tai`** calendar.
For any given instant, the **`tai`** datetime is ahead of the **`utc`** datetime, where "ahead" means the same as it does when describing a timezone to the east as being ahead of one to the west.
Moreover, a given datetime in the **`tai`** calendar represents an instant that is later than the same datetime in the **`utc`** calendar.

> The difference between the two datetimes (**`tai`** minus **`utc`**) for a given instant of time is the net number of leap seconds introduced since 1958-01-01.

Author: I've made this change too. Of course it's debatable, but I think it's helpful since you never know where a reader will go first. You previously objected to the statement earlier in the calendar section: "The net number of leap seconds added to UTC between 1958-1-1 and 2025-1-1 is 37." Since this is a paragraph _only_ about UTC (TAI is not involved), and we're now starting UTC in 1972, I've changed it to, "The net number of leap seconds added to UTC between 1972-1-1 and 2025-1-1 is 27." I hope that's OK.

A non-zero time zone offset is not allowed in this calendar.

**`noleap`** or **`365_day`**:: A calendar with no leap years, i.e., all years are 365 days long, and there are no leap seconds.

**`all_leap`** or **`366_day`**:: A calendar in which every year is a leap year, i.e., all years are 366 days long, and there are no leap seconds.

**`360_day`**:: A calendar in which all years are 360 days, and divided into 30 day months, and there are no leap seconds.

**`none`**:: To be used when there is no annual cycle.
See <<none-calendar>>.

Any other value may be given to the **`calendar`** attribute to describe an explicitly defined calendar.
See <<explicit-calendar>>.

[[leap-seconds, Section 4.4.4, "Leap Seconds"]]
==== Leap Seconds

> This section describes how to deal properly with leap seconds.

Reviewer 2: I strongly oppose to any language suggesting that CF has a solution to "how to deal properly with leap seconds". This is an identified  problem that currently is without an accepted solution. There are several alternative technical solutions cf. this presentation https://www.bipm.org/documents/20126/52354616/CCTF_UTC_presentation.pdf/0a8f0954-7c00-9142-8da2-f4221c7e0269 (in particular I draw your attention to viewgraph 12), and this paper https://doi.org/10.1088/1681-7575/ac9da5.

Most people ignore the existence of leap seconds, including many data producers and the CF standard before version 1.12.

> __If you are producing real-world datasets with datetimes in UTC, and if it's important for your time coordinates to be accurate to the second, the **`utc`** calendar is recommended.__

Reviewer 2: This will requires a longer response-- see email.

See <<leap-seconds>>.

The CF standard follows UDUNITS (<<units>>) in the definition of the acceptable
units of measure for time.
The most commonly used of these units strings (and their abbreviations) are **`day`** (**`d`**), **`hour`** (**`hr`**, **`h`**), **`minute`** (**`min`**) and **`second`** (**`sec`**, **`s`**).
Plural forms are also acceptable.
Any unit may optionally have one of the decimal prefixes for multiples and submultiples (<<table-supported-units, Table 3.1>>), but these are uncommon except for submultiples of **`second`** e.g. **`millisecond`** or **`ms`**.

UDUNITS defines a **`year`** to be exactly 365.242198781 days (the interval between 2 successive passages of the sun through vernal equinox).
__It is not a calendar year.__ UDUNITS defines a  **`month`** to be exactly **`year/12`**, which is __not a calendar month__.
We recommend that **`year`** and **`month`** should not be used, because of the potential for mistakes and confusion.

UDUNITS defines a **`minute`** as 60 **`seconds`**, an **`hour`** as 3600 **`seconds`** and a **`day`** as 86400 **`seconds`**.

> These are not calendar units.

Reviewer 2: Following the suggested change to the previous line I suggest to delete this line. From the context here it is not absolutely clear what a "calendar unit" is, and when/where they might be different from what is common knowledge. The previous line provide all the precision that is necessary at this point in the text.

Author: I've deleted "These are not calendar units", and moved the "fixed units of measure" phrase, which we already had, to replace it.

When a leap second is inserted into UTC, the minute, hour and day affected differ by one second from their usual durations according to clock time, but the units of **`minute`**, **`hour`** and **`day`** do not; they are fixed units of measure.
To avoid mistakes and confusion, we therefore recommend that these units should not be used in the **`utc`** calendar; instead, we recommend the **`second`** as the unit for time coordinates in that calendar.
See also <<calendar>> and <<leap-seconds>>.

UDUNITS permits a number of alternatives to the word **`since`** in the units of time coordinates.
All the alternatives have exactly the same meaning in UDUNITS.
For compatibility with other software, CF strongly recommends that **`since`** should be used.

The reference datetime string (appearing after the identifier **`since`**) is required.
It must include the date, which may be followed by time or time zone offset or both.

> Its format is __y__-__m__-__d__ [__H__:__M__:__S__] [__Z__], where [...] indicates an optional element, as follows:

Reviewer 2: ```suggestion
It must include the date, which may be followed by time and optionally a time zone offset.
Its format is __y__-__m__-__d__ [__H__:__M__:__S__ [__Z__]], where [...] indicates an optional element, as follows:
```
UDUNITS2 does not follow its own (rather complex and sketchy/incomplete) grammar when it comes to handling of the reference time. And CF should not simply accept what UDUNITS2 does and the formatting rules behind that. Without going into details, and I do *not* have the full picture, I very strongly argue that CF must not attempt to describe what UDUNITS2 is *actually* doing; that would amount to writing a new documentation of that part of UDUNITS2, including updating the grammar. Below are some examples, from which I conclude that the best we can do is to allow a time zone *only if the time is explicitly stated* (partial time, i.e. only hour is OK), which is what the suggested change above aims to achieve.
```bash
$ udunits2 -H "days since 1990-1-1 00:00:00" -W ""
    (86400 s) @ 19900101T000000.00000000 UTC             # OK
$ udunits2 -H "days since 1990-1-1 00:00:00-10" -W ""
    (86400 s) @ 19900101T100000.00000000 UTC             # OK
$ udunits2 -H "days since 1990-1-1 00:00:00 -10" -W ""
    (86400 s) @ 19900101T100000.00000000 UTC             # OK
$ udunits2 -H "days since 1990-1-1 00:00:00 +10" -W ""
    (86400 s) @ 19891231T140000.00000000 UTC             # OK
$ udunits2 -H "days since 1990-1-1 00:00:00+10" -W ""
    (86400 s) @ 19891231T140000.00000000 UTC             # OK
$ udunits2 -H "days since 1990-1-1 00:00:00Z" -W ""
    (86400 s) @ 19900101T000000.00000000 UTC             # OK
$ udunits2 -H "days since 1990-1-1 00:00:00 Z" -W ""
    (86400 s) @ 19900101T000000.00000000 UTC             # OK
$ udunits2 -H "days since 1990-1-1 00:00-10" -W ""
    (86400 s) @ 19900101T100000.00000000 UTC             # OK
$ udunits2 -H "days since 1990-1-1 00-10" -W ""
    (86400 s) @ 19900101T100000.00000000 UTC             # OK
$ udunits2 -H "days since 1990-1-1 -10" -W "" 
    (86400 s) @ 19891231T230000.00000000 UTC             # NOT OK
$ udunits2 -H "days since 1990-1-1-10" -W ""
    (86400 s) @ 19891231T230000.00000000 UTC             # NOT OK  
$ udunits2 -H "days since 1990-1-1 Z" -W ""
    (86400 s) @ 19900101T000000.00000000 UTC             # OK
$ udunits2 -H "days since 1990-1-1Z" -W ""
    (86400 s) @ 19900101T000000.00000000 UTC             # OK
$ udunits2 -H "days since 1990-1-1+10" -W ""
    (86400 s) @ 19900101T010000.00000000 UTC             # NOT OK
$ udunits2 -H "days since 1990-1-1+13" -W ""
    (86400 s) @ 19900101T010300.00000000 UTC             # NOT OK
$ udunits2 -H "days since 1990-1-1+11" -W ""
    (86400 s) @ 19900101T010100.00000000 UTC             # NOT OK

```

Author: Thanks for this investigation. I agree that it is wrong, and I also get the wrong answer when omitting the time:
```
You have: days since 1990-1-1 -10
You want: days since 1990-1-1 10:00
    1 days since 1990-1-1 -10 = 0.541667 (days since 1990-1-1 10:00)
    x/(days since 1990-1-1 10:00) = (x/(days since 1990-1-1 -10)) - 0.458333
You have: days since 1990-1-1 00:00 -10
You want: days since 1990-1-1 10:00
    1 days since 1990-1-1 00:00 -10 = 1 (days since 1990-1-1 10:00)
    x/(days since 1990-1-1 10:00) = (x/(days since 1990-1-1 00:00 -10))
```
where the second answer is correct and the first should mean the same.

I agree with and have adopted your suggestion that we had better require the time to be stated if the time zone is included, which makes sense really. I also agree with the general point that CF needs to state explicitly what the units syntax is, as we have previously discussed. That should be the subject of another issue.

* __y__ is year, __m__ month, __d__ day, __H__ hour and __M__ minute, which are all integers of one or more digits, and __y__ may be prefixed with a sign (but note that some CF calendars do not permit negative years; see <<calendar>>),

* __S__ is second, which may be integer or floating point (see <<leap-seconds>> regarding __S__,

* __Z__ is the time zone offset.
__Z__ is __not__ a time zone name or acronym; it is an interval of time.

The default for time zone offset __Z__ is zero, which may also be explicitly indicated in any of the numeric formats for __Z__ defined below, or by the letter **`Z`**, sometimes referred to as "Zulu Time".

> We suggest that a zero offset be stated explicitly to avoid confusion in any situation where omitting it might be misunderstood as indicating local time.

Author: Fair point. I have replaced _Z_ with _T_. Is that OK? I'd rather not use more than one letter.

Thus, **`units = "days since 1990-1-1"`** means the same as **`units = "days since 1990-1-1 00:00:00"`**.

We recommend that a non-zero time zone offset should not be specified, because it is easy to make mistakes about the sign of the offset and allowance for daylight-saving/summer time.
A non-zero offset is __not allowed__ in the **`utc`** and **`tai`** calendars (see <<calendar>>).

Other than **`Z`** for zero offset, the time zone offset __Z__ must be in one of the following four numeric formats, each of which may optionally be prefixed by a positive sign **`+`** or negative sign **`-`**, the default sign being positive:

* __H__, the hour alone, of one or two digits e.g. **`-6`**, **`2`**, **`+11`**, which is sufficient for many time zones.

* __H__:__M__, where __H__ is hour and __M__ minute, each of one or two digits, and separated by a colon **`:`**, e.g. **`5:30`**.

* four digits, of which the first pair are the hours and the second the minutes e.g. **`0530`**.

* three digits, of which the first is the hour (0--9) e.g. **`530`**.

Spaces are not allowed in the numeric formats.

> If __Z__ is **`Z`** or begins with **`+`** or **`-`**, the space before __Z__ may be omitted.

Author: _Z_ has been replaced with _T_ everywhere.

For example, **`seconds since 1992-10-8 15:15:42.5 -6:00`** indicates seconds since October 8th, 1992 at 3 hours, 15 minutes and 42.5 seconds in the afternoon, in a time zone where the datetime is six hours behind the default.
Subtracting the time zone offset from a given datetime converts it to the equivalent datetime with zero time zone offset e.g. **`1989-12-31 18:00:00 -6`** identifies the same instant as **`1990-1-1 0:0:0`**.

[[calendar, Section 4.4.3, "Calendar"]]
==== Calendar

The calendar defines the set of valid datetimes and their order.
Note that the CF meaning of "calendar" refers to datetimes, not to dates alone.
Datetimes which are not permitted in a given calendar are prohibited both in the time coordinate values and in the reference datetime string in the **`units`**.
It is recommended that the calendar be specified by the **`calendar`** attribute of the time coordinate variable (rather than relying on the default).
The values currently defined for **`calendar`** are listed below.

> Because the calendars have different sets of valid dates, and different treatments of leap seconds (see below in this section, and <<leap-seconds>>), a given time coordinate value with given **`units`** can represent different datetimes in different calendars; conversely, a given datetime is represented by different time coordinate values in different calendars.

Author: Fair enough; "datetimes" is better than "dates", and leap seconds have not yet been mentioned in this section.

** three digits, of which the first is the hour (0--9) e.g. **`530`**.

For example, **`seconds since 1992-10-8 15:15:42.5 -6:00`** indicates seconds since October 8th, 1992 at 3 hours, 15 minutes and 42.5 seconds in the afternoon, in a time zone where the datetime is six hours behind the default.

> To translate a given time coordinate to the time zone of the reference datetime, the time zone offset should be added to the time coordinate before converting it to a datetime.

Reviewer: I don't understand this ... a time coordinate is, by definition, an elapsed time since the reference datetime, so what are we converting it to?

Author: I don't know what it was supposed to mean. I have deleted it.

> Because the calendars have different sets of valid dates, and different treatments of leap seconds (see below in this section, and <<leap-seconds>>), a given time coordinate value with given **`units`** can represent different datetimes in different calendars; conversely, a given datetime is represented by different time coordinate values in different calendars.

Author: I've deleted the parenthesis as you suggest. I haven't changed "datetime" to "instant", although what you say is correct, because the "conversely" is intended to be exactly the reverse.

> Moreover, in different calendars a given datetime can identify a different instant in the continuous physical dimension of time.

Reviewer 2: The current text (and earlier versions, too, I believe) begs the potentially difficult question "what calendar does the reference datetime itself represent?" I do not think that has be clarified? To my understanding it is the standard or utc calendar.

> The lengths of the months in the Gregorian calendar are used in all calendars except **`360_day`**, **`none`** (see <<none-calendar>>) and explicitly defined calendars (see <<explicit-calendar>>).
> The calendars differ in their treatment of leap years (when there are 29 days in February instead of 28).

Reviewer 4: ```suggestion

```
Delete sentences as not relevant.

In response to slight variations in the Earth's rotation speed, positive or negative leap seconds are inserted in order to keep UTC close to mean solar time at 0 `degrees_east` i.e. the time zone with the default (zero) time zone offset in UDUNITS and CF (see <<time-coordinate-units>>).
When a single positive leap second is introduced at the end of a minute, that minute contains 61 seconds.
The net number of leap seconds added to UTC between 1958-1-1 and 2025-1-1 is 37.
Leap seconds should be considered when processing datetimes and time-coordinates in the **`standard`** and **`utc`** calendars, which treat them differently (see <<leap-seconds>>).
The other CF calendars do not allow for leap seconds.

In the **`julian`** and the default **`standard`** calendar, dates in years before year 0 (i.e. before 0-1-1 0:0:0) are not allowed, and the year in the reference datetime of the units must not be negative.
In these calendars, year zero has a special use to indicate a climatology (see <<climatological-statistics>>), but this use of year zero is deprecated.

> In other calendars, year 0 is the year before year 1, and negative years are allowed.

Reviewer 4: There is a difference between the Julian calendar, in which there is no year 0, and the CF `julian` calendar in which year 0 has previously been used to denote climatological statistics. In either case, my suggestion is to delete the sentences here and add better language in the sub-section of the `julian` calendar.

The possible values for the **`calendar`** attribute are as follows:

> **`standard`**:: The historical Gregorian--Julian calendar as defined by UDUNITS.

Reviewer 4: ```suggestion
**`standard`**:: The Gregorian--Julian calendar as defined by UDUNITS.
```

Author: Done

> This is the default.

Reviewer 4: ```suggestion
This is the default calendar when a `calendar` attribute is not specified in the time coordinate variable (except when attribute `month_lengths` is present; see section on <<explicit_calendar>>).
```

> In the **`standard`** calendar, 1582-10-15 0:0:0 is exactly 1 day later than 1582-10-4 0:0:0.
> Therefore datetimes in the range from (and including) 1582-10-5 0:0:0 until (but excluding) 1582-10-15 0:0:0 are invalid, and must not be used as reference in **`units`**.

Reviewer 4: ```suggestion
In the **`standard`** calendar, 1582-10-15 immediately follows 1582-10-04.
Therefore datetimes in the range from (and including) 1582-10-05 until (but excluding) 1582-10-15 are invalid, and must not be used as reference in **`units`**.
```

It is recommended that a reference datetime before the discontinuity should not be used for datetimes after the discontinuity, and vice-versa.
+

> __The **`standard`** calendar is used both for (mostly observational) data with real-world datetimes, including leap seconds (after 1972), and for data not including leap seconds__, generated by models according to the Gregorian calendar.

Reviewer: Not sure if this emphsis is required, given that the following sentence is emphasised.
```suggestion
The **`standard`** calendar is used both for (mostly observational) data with real-world datetimes, including leap seconds (after 1972); and for data not including leap seconds, generated by models according to the Gregorian calendar.
```

Author: I wanted to emphasise first that the standard calendar is used (even though we may not like it) both for data with leap seconds and for data without leap seconds. The next emphasis is the caution, which is a consequence of it.

Therefore, __when using data written with the **`standard`** calendar, you should regard datetimes and time coordinates as uncertain by a number of seconds (but less than a minute),__ because leap seconds may have been counted when you expected them to have been ignored, or ignored when you expected them to have been counted.
For existing datasets, this uncertainty can be eliminated only if you have separate information about how leap seconds were treated.
For producing new datasets, where precision to the second is important, it is recommended to use the **`utc`** calendar for data that follows the real-world calendar according to UTC including leap seconds, and the **`proleptic_gregorian`** calendar for model-generated data that follows the Gregorian calendar without leap seconds.
See also <<leap-seconds>>.

> **`proleptic_gregorian`**:: A calendar with the Gregorian rules for leap years extended to dates before 1582-10-15.

Reviewer 4: ```suggestion
**`proleptic_gregorian`**:: A Gregorian calendar continuously extended to dates before 1582-10-15. Days in the range 1582-10-05 to 1582-10-14 exist, as well as year 0 and negative years. 
```

Author: I think my version of the first sentence is more accurate. There is only one Gregorian calendar in world history, which was implemented on 1582-10-15 (as represented by UDUNITS). We aren't extending the Gregorian calendar; it's a new invention which uses the same algorithm to cover the years before the Gregorian calendar began.

My next sentence is "All dates consistent with these rules are allowed, both before and after 1582-10-15 0:0:0." In view of your second suggested sentence, I've changed mine to "All dates consistent with these rules are allowed, including those in the range 1582-10-05 to 1582-10-14." There's already a general statement (in the paragraph about year zero) saying "In other calendars [apart from `standard` and `julian`], year 0 is the year before year 1, and negative years are allowed."

> All dates consistent with these rules are allowed, both before and after 1582-10-15 0:0:0.
> For model-generated data following the Gregorian calendar without leap seconds, the **`proleptic_gregorian`** calendar is recommended, in preference to the **`standard`** (or default) calendar.
> See also <<leap-seconds>>.

Reviewer 4: ```suggestion

```

Delete these sentences and describe in a section on interpreting and selecting calendars

**`julian`**:: Julian calendar, in which a year is a leap year if it is divisible by 4, even if it is also divisible by 100.
Year 1 AD or CE in the **`julian`** calendar is also year 1 of the **`standard`** calendar.
Negative years are invalid in time coordinates and reference datetimes in the **`julian`** calendar, and leap seconds are not allowed.

**`utc`**:: A Gregorian calendar __with__ leap seconds as prescribed by UTC.

> Datetimes before 1972-01-01 0:0:0 are not allowed in this calendar.

Reviewer 4: ```suggestion
The current definition of UTC came into force on  1972-01-01 and is based on TAI with a difference of 10 seconds at that time: 1972-01-01 00:00:00 UTC denotes the same instant as 1972-01-01 00:00:10 TAI. Since its start, 27 leap seconds have been inserted in UTC meaning that, at the Tim elf publication of this document, UTC is 37 seconds behind TAI.
```

> Datetimes in the future are not allowed in this calendar, because it is unknown when future leap seconds will occur.

Reviewer 4: ```suggestion
Datetimes before 1972-01-01 are not allowed in this calendar. Datetimes in the future are not allowed in this calendar either, because it is unknown when future leap seconds will be inserted. A non-zero time zone offset is not allowed in this calendar.
```

> In this calendar, the second is the only recommended unit of measure in the **`units`** of the time coordinate variable (denoted by any string allowed in UDUNITS, optionally with one of the prefixes of <<table-supported-units, Table 3.1>>).

Reviewer 4: ```suggestion
In this calendar, it is recommended to use the second as the unit of measure in the **`units`** attribute of the time coordinate variable.
```

> When a datetime is converted to a time coordinate value or vice-versa in this calendar, any leap seconds (positive or negative) must be counted that occurred in the interval between the datetime and the reference datetime in the **`units`**.

Reviewer 4: ```suggestion
When a datetime is converted to a time coordinate value or vice-versa in this calendar, any leap seconds (positive or negative) must be counted that occurred in the interval between the datetime and the reference datetime in the **`units`** attribute.
```

Author: I don't think we need the word "attribute". It is often omitted in the CF document when it's already clear we're talking about an attribute. The **`font`** indicates this, and also that if we meant just plain "units" (of measure) there would be no reference datetime.

> For any given instant, the **`utc`** datetime is behind the **`tai`** datetime, where "behind" means the same as it does when describing a timezone to the west as being behind one to the east.
> Moreover, a given datetime in the **`utc`** calendar represents an instant that is earlier than the same datetime in the **`tai`** calendar.
> The difference between the two datetimes for a given instant of time is the net number of leap seconds introduced since 1958-01-01.

Reviewer 4: ```suggestion

```
This is not the place to discuss the differences between both calendars

> For observational or model-generated data that follows the Gregorian calendar and UTC (after 1972) including leap seconds, where precision to the second is important, the **`utc`** calendar is recommended, in preference to the **`standard`** (or default) calendar.

Reviewer 4: ```suggestion

```
As above

See also <<leap-seconds>>.

> **`tai`**:: A Gregorian calendar __without__ leap seconds that is based on International Atomic Time (TAI).

Reviewer 4: ```suggestion
**`tai`**:: A Gregorian calendar that is based on International Atomic Time (TAI).
```
By definition there are no leap seconds so no need to emphasise that

> Datetimes before 1958-01-01 0:0:0 are not allowed in this calendar.

Reviewer 4: ```suggestion
TAI was first introduced on 1958-01-01 and earlier datetimes are thus not allowed in this calendar. A non-zero time zone offset is not allowed in this calendar.
```

> For any given instant, the **`tai`** datetime is ahead of the **`utc`** datetime, where "ahead" means the same as it does when describing a timezone to the east as being ahead of one to the west.
> Moreover, a given datetime in the **`tai`** calendar represents an instant that is later than the same datetime in the **`utc`** calendar.
> The difference between the two datetimes for a given instant of time is the net number of leap seconds introduced since 1958-01-01.

Reviewer 4: ```suggestion

```
Not relevant here

Author: I think it's useful to spell this out in detail, because it's easy to get confused about.

__The **`standard`** calendar is used both for (mostly observational) data with real-world datetimes, including leap seconds (after 1972), and for data not including leap seconds__, generated by models according to the Gregorian calendar.
Therefore, __when using data written with the **`standard`** calendar, you should regard datetimes and time coordinates as uncertain by a number of seconds (but less than a minute),__ because leap seconds may have been counted when you expected them to have been ignored, or ignored when you expected them to have been counted.
For existing datasets, this uncertainty can be eliminated only if you have separate information about how leap seconds were treated.

> For producing new datasets, where precision to the second is important, it is recommended to use the **`utc`** calendar for data that follows the real-world calendar according to UTC including leap seconds, and the **`proleptic_gregorian`** calendar for model-generated data that follows the Gregorian calendar without leap seconds.

Reviewer 5: maybe a couple commas?

For producing new datasets, where precision to the second is important, it is recommended to use the **`utc`** calendar for data that follows the real-world calendar, according to UTC, including leap seconds, and the **`proleptic_gregorian`** calendar for model-generated data that follows the Gregorian calendar without leap seconds.

When a datetime is converted to a time coordinate or vice-versa in the **`utc`** calendar, any leap seconds must be counted that occurred between the instant concerned and the reference datetime in the **`units`** attribute.

For example, a leap second was added to UTC at the end of 2016.
Therefore if **`calendar="utc"`** and **`units="seconds since 2016-12-31 23:59:58"`**, a value of 4 for the time coordinate represents the datetime 2017-01-01 00:00:01, because four seconds had elapsed by that instant since 2016-12-31 23:59:58 (seconds ending at 2016-12-31 23:59:59,  2016-12-31 23:59:60,  2017-01-01 00:00:00,  2017-01-01 00:00:01).
In the **`utc`** calendar, two instants with the same time of day on different days, which would always be separated by a multiple of 86400 seconds if there were no leap seconds, will have a few more seconds between them if positive leap seconds intervene.
For example, with respect to the the reference datetime in **`units="seconds since 2016-12-31 23:59:58"`**, the datetime 2017-01-01 23:59:58, which is one calendar day later, is represented by a time coordinate of 86401.

> In all calendars except **`utc`** and **`standard`** calendar, datetimes within leap seconds are not valid, and reference datetimes in the **`units`** attribute must not contain seconds equal to or greater than 60.

Reviewer: ```suggestion
In all calendars except **`utc`** and **`standard`**, datetimes within leap seconds are not valid, and reference datetimes in the **`units`** attribute must not contain seconds equal to or greater than 60.
```

__If you are producing model-generated datasets with datetimes that follow the Gregorian calendar but do not include leap seconds, the **`proleptic_gregorian`** calendar is recommended.__
The **`standard`** calendar may alternatively be used, but is not recommended because of its uncertainty about leap seconds, described in the paragraphs after this one.
Since the **`proleptic_gregorian`** calendar includes no leap seconds, the time coordinate for a given datetime and **`units`** may differ between this calendar and **`utc`**.
For example, if **`calendar="proleptic_gregorian"`** and **`units="seconds since 2016-12-31 23:59:58"`**, the datetime 2017-01-01 00:00:01 is represented by a time coordinate of 3, because the UTC leap second (from 2016-12-31 23:59:60 to 2017-01-01 00:00:00) does not exist in the **`proleptic_gregorian`** calendar, whereas the time coordinate is 4 in **`utc`** with the same **`units`** attribute.
In the **`proleptic_gregorian`** calendar, the time coordinate for 2017-01-01 00:00:58 is 60, and for 2017-01-01 23:59:58 is 86400.

> __The **`standard`** calendar is used both for real-world data, and for data generated by models that follow the Gregorian calendar and do not include leap seconds in the timeline.__

Reviewer: Not emphasised? (similarly to my  suggestion on the similar sentence in the `standard` calendar description)
 ```suggestion
The **`standard`** calendar is used both for real-world data, and for data generated by models that follow the Gregorian calendar and do not include leap seconds in the timeline.
```
